# Data_preprocessing_and_adding_features_set_1 , Here we have loaded data and tried to understand it and added some intial stage features

In [ ]:
import numpy as np
import pandas as pd
import networkx as nx

In [2]:
df = pd.read_csv('Data\combined_data.csv', header = None)

df here contains entire dataset as edge list 

In [3]:
#converting all values to int 
df = df.astype(int)
df

0        1
0          687794  2132718
1          687794   364024
2          687794  3273020
3          687794  4589838
4          687794  4507085
...           ...      ...
12903547  3276950  3604072
12903548  3276950  3018756
12903549  3276950   362031
12903550  3276950  1008178
12903551  3276950   391199

[12903552 rows x 2 columns]

In [4]:
headers = ['Form', 'To']
df.columns = headers
df

Form       To
0          687794  2132718
1          687794   364024
2          687794  3273020
3          687794  4589838
4          687794  4507085
...           ...      ...
12903547  3276950  3604072
12903548  3276950  3018756
12903549  3276950   362031
12903550  3276950  1008178
12903551  3276950   391199

[12903552 rows x 2 columns]

In [5]:
# finding min and max here
from_column = df['Form']
print(from_column)

min_value = from_column[0:].min()
max_value = from_column[0:].max()
print(min_value)
print(max_value)

0            687794
1            687794
2            687794
3            687794
4            687794
             ...   
12903547    3276950
12903548    3276950
12903549    3276950
12903550    3276950
12903551    3276950
Name: Form, Length: 12903552, dtype: int32
500
4866749


In [25]:
#Here i am sampling 25,000 points randomly from above dataset
sampled_positive_edge = df.sample(n=90000, random_state=42)
sampled_positive_edge

Form       To
5363684   4862306  3513873
12097305   629760  4763045
3009767   3086336  1009858
9058126   1090555  4388892
5554601    307789  4761599
...           ...      ...
4583458    630167   411109
3509       314520  4207564
8779359   2033249   749138
4100287   2943850   940499
7670537   1694805  2225791

[90000 rows x 2 columns]

In [26]:
type(sampled_positive_edge['Form'])

pandas.core.series.Series

In [27]:
#number of unique ele in sampled_positive_edge
num_unique = len(sampled_positive_edge['Form'].unique())
print(num_unique)

12215


The above uniqueness seems fine as of now

In [28]:
test_data = pd.read_csv('real_test_data.csv')
test_data

Id     Form       To
0        1  3360982  4457271
1        2  4761876  4698439
2        3  4198430  3615486
3        4  2945770   747948
4        5  3950088  3360335
...    ...      ...      ...
1995  1996  3116985  2415415
1996  1997   265459   842701
1997  1998  2547139  4332939
1998  1999  2366264  1815056
1999  2000  2735577  1232247

[2000 rows x 3 columns]

In [29]:
set_col0 = set(sampled_positive_edge['Form'])
set_col1 = set(test_data['Form'])
common_elements = set_col0.intersection(set_col1)
num_common_elements = len(common_elements)
print(num_common_elements)

1578


Till Here it seems fine as half of test data is there in train data i havent included the false edges till now 

In [91]:
import random
import csv
import pickle
import os


    
edges = dict()

for edge in df:
        edges[(edge[0], edge[1])] = 1

        
missing_edges = set([])

while (len(missing_edges)<25000):
        a=random.randint(500, 4866749)
        b=random.randint(1, 4866749)
        tmp = edges.get((a,b),-1)
        if tmp == -1 and a!=b:
            try:
                if nx.shortest_path_length(g,source=a,target=b) > 2: 

                    missing_edges.add((a,b))
                else:
                    continue  
            except:  
                    missing_edges.add((a,b))              
        else:
            continue

In [92]:
len(missing_edges)

25000

In [49]:
missing_edges

{(2723836, 2019352),
 (624123, 4164077),
 (2965284, 683171),
 (2120676, 173517),
 (516995, 564114),
 (229295, 3924089),
 (4157406, 1713939),
 (4187180, 410003),
 (3937057, 3690713),
 (3820837, 2979591),
 (308572, 683993),
 (2472714, 3194407),
 (2504214, 4696755),
 (773581, 4122263),
 (512401, 2128413),
 (803619, 2109933),
 (1303735, 1114539),
 (4576246, 568725),
 (4649060, 876375),
 (2459738, 1451648),
 (1990720, 3445003),
 (4525837, 3836578),
 (2356583, 2001766),
 (4284898, 4797276),
 (1305247, 2071829),
 (2525057, 4713965),
 (3901379, 447192),
 (1322748, 771875),
 (2215147, 4311575),
 (2969546, 1371792),
 (1931002, 1815326),
 (4252903, 2461030),
 (4666797, 342077),
 (4182372, 3501183),
 (981530, 1094310),
 (295930, 141983),
 (3046233, 1824581),
 (4151865, 3291293),
 (606459, 3149602),
 (195115, 2893760),
 (551350, 4411226),
 (4074717, 1772187),
 (74335, 1549747),
 (164443, 1583795),
 (1089607, 3490088),
 (4010890, 104796),
 (239137, 1206698),
 (3190252, 3066835),
 (3114286, 4048614),

In [93]:
missing_edges_list = list(missing_edges)
missing_edges_list


[(1412960, 159984),
 (450019, 3048705),
 (2341943, 2879266),
 (3936197, 4732550),
 (258980, 970828),
 (1100634, 3150332),
 (2638614, 550345),
 (4173010, 1967164),
 (4223388, 3300744),
 (1617787, 2681260),
 (2077593, 1164001),
 (243557, 344698),
 (3476012, 388278),
 (466126, 2775324),
 (4490784, 1182238),
 (3874502, 3488730),
 (3328986, 1177421),
 (3525998, 2020210),
 (1830816, 4057495),
 (3715473, 587074),
 (1796021, 3649987),
 (25884, 3786505),
 (1492855, 1210951),
 (1137008, 4797754),
 (3455403, 3100537),
 (3062171, 4245703),
 (4238442, 4104433),
 (3940810, 4821071),
 (296453, 3707054),
 (56681, 2709001),
 (2316425, 3648854),
 (407534, 760526),
 (2766932, 2106402),
 (2987365, 2716763),
 (4713613, 1720184),
 (1527155, 3465370),
 (3351784, 762410),
 (3376116, 385573),
 (1004520, 351331),
 (928752, 1198406),
 (4702121, 3265151),
 (2261138, 3465548),
 (1485472, 1920897),
 (3285594, 1214979),
 (562636, 1912573),
 (2148763, 2005750),
 (3020674, 1073778),
 (2124113, 975584),
 (1562283, 2846

In [94]:
type(missing_edges_list[0])

tuple

In [95]:
#column_names = ['ID', 'Name']

# Create a Pandas DataFrame
df_false_ed = pd.DataFrame(missing_edges_list)

In [96]:
df_false_ed

0        1
0      1412960   159984
1       450019  3048705
2      2341943  2879266
3      3936197  4732550
4       258980   970828
...        ...      ...
24995   918964   186086
24996  2561631  2260842
24997  4741601  4778408
24998  1692402  2786562
24999  2931013  1568451

[25000 rows x 2 columns]

In [98]:
headers = ['Form', 'To']
df_false_ed.columns = headers
df_false_ed

Form       To
0      1412960   159984
1       450019  3048705
2      2341943  2879266
3      3936197  4732550
4       258980   970828
...        ...      ...
24995   918964   186086
24996  2561631  2260842
24997  4741601  4778408
24998  1692402  2786562
24999  2931013  1568451

[25000 rows x 2 columns]

In [97]:


sampled_positive_edge

Form       To
5363684   4862306  3513873
12097305   629760  4763045
3009767   3086336  1009858
9058126   1090555  4388892
5554601    307789  4761599
...           ...      ...
1931898   4177982  2652529
8497529    305271  1110562
3922966   1522576  4630277
10484541  1229957   349493
9897533    690509   534148

[25000 rows x 2 columns]

In [57]:
set_col0 = set(df_false_ed[0])
set_col1 = set(test_data['From'])
common_elements = set_col0.intersection(set_col1)
num_common_elements = len(common_elements)
print(num_common_elements)

7


It might create somr problem here but lets go ahead with this

## Now i will be adding Features Here

## Feature_1_is_in_degree

In [99]:
df

Form       To
0          687794  2132718
1          687794   364024
2          687794  3273020
3          687794  4589838
4          687794  4507085
...           ...      ...
12903547  3276950  3604072
12903548  3276950  3018756
12903549  3276950   362031
12903550  3276950  1008178
12903551  3276950   391199

[12903552 rows x 2 columns]

#Here doing featureization on entire data

In [60]:
G_complete_train_data = nx.from_pandas_edgelist(df, source='Form', target='To', create_using=nx.DiGraph())

In [62]:
# Calculate indegree and outdegree for each node
indegree = dict(G_complete_train_data.in_degree())
outdegree = dict(G_complete_train_data.out_degree())

In [100]:
print(len(indegree))
len(outdegree)

2480801


2480801

In [66]:
src_indegree_df = pd.DataFrame(list(indegree.items()), columns=['Node', 'src_Indegree'])
src_outdegree_df = pd.DataFrame(list(outdegree.items()), columns=['Node', 'src_Outdegree'])

In [67]:
src_indegree_df 

Node  src_Indegree
0         687794           135
1        2132718            37
2         364024            75
3        3273020            27
4        4589838           102
...          ...           ...
2480796   827099             1
2480797  4006004             1
2480798  3988075             1
2480799  4436196             1
2480800  1637552             1

[2480801 rows x 2 columns]

In [71]:
from_column = src_indegree_df ['src_Indegree']
print(from_column)

min_value = from_column[0:].min()
max_value = from_column[0:].max()
print(min_value)
print(max_value)

0          135
1           37
2           75
3           27
4          102
          ... 
2480796      1
2480797      1
2480798      1
2480799      1
2480800      1
Name: src_Indegree, Length: 2480801, dtype: int64
1
4639


In [68]:
src_outdegree_df

Node  src_Outdegree
0         687794            143
1        2132718             20
2         364024             79
3        3273020             26
4        4589838            222
...          ...            ...
2480796   827099              0
2480797  4006004              0
2480798  3988075              0
2480799  4436196              0
2480800  1637552              0

[2480801 rows x 2 columns]

In [103]:
from_column = src_outdegree_df['src_Outdegree']
print(from_column)

min_value = from_column[0:].min()
max_value = from_column[0:].max()
print(min_value)
print(max_value)

0          143
1           20
2           79
3           26
4          222
          ... 
2480796      0
2480797      0
2480798      0
2480799      0
2480800      0
Name: src_Outdegree, Length: 2480801, dtype: int64
0
7999


In [104]:
combined_data_with_attributes = pd.merge(sampled_positive_edge, src_indegree_df , left_on='Form', right_on='Node', how='left')
combined_data_with_attributes = pd.merge(combined_data_with_attributes, src_outdegree_df, left_on='Form', right_on='Node', how='left')
                    

In [105]:
combined_data_with_attributes

Form       To   Node_x  src_Indegree   Node_y  src_Outdegree
0      4862306  3513873  4862306           291  4862306           7998
1       629760  4763045   629760           144   629760           1351
2      3086336  1009858  3086336           152  3086336           2395
3      1090555  4388892  1090555            50  1090555            295
4       307789  4761599   307789           161   307789            664
...        ...      ...      ...           ...      ...            ...
24995  4177982  2652529  4177982           151  4177982           7999
24996   305271  1110562   305271           430   305271            435
24997  1522576  4630277  1522576           629  1522576           7998
24998  1229957   349493  1229957           696  1229957           7999
24999   690509   534148   690509           103   690509           3070

[25000 rows x 6 columns]

In [106]:
columns_to_drop = ['Node_x', 'Node_y']
combined_data_with_attributes = combined_data_with_attributes.drop(columns=columns_to_drop)

In [107]:
combined_data_with_attributes

Form       To  src_Indegree  src_Outdegree
0      4862306  3513873           291           7998
1       629760  4763045           144           1351
2      3086336  1009858           152           2395
3      1090555  4388892            50            295
4       307789  4761599           161            664
...        ...      ...           ...            ...
24995  4177982  2652529           151           7999
24996   305271  1110562           430            435
24997  1522576  4630277           629           7998
24998  1229957   349493           696           7999
24999   690509   534148           103           3070

[25000 rows x 4 columns]

In [111]:
average_src_Indegree = combined_data_with_attributes['src_Indegree'].mean()
average_src_Outdegree = combined_data_with_attributes['src_Outdegree'].mean()
print(average_src_Indegree)
print(average_src_Outdegree)
med_in = combined_data_with_attributes['src_Indegree'].median()
med_out = combined_data_with_attributes['src_Outdegree'].median()
print(med_in)
print(med_out)

252.31388
3877.08676
132.0
2951.0


In [112]:
in_degrees_dest = G_complete_train_data.in_degree(combined_data_with_attributes['To'])
lst = list(in_degrees_dest)
type(lst[0])
second_values = [t[1] for t in lst]
second_values
combined_data_with_attributes['In_Degree_dest'] = second_values
combined_data_with_attributes

Form       To  src_Indegree  src_Outdegree  In_Degree_dest
0      4862306  3513873           291           7998              11
1       629760  4763045           144           1351              23
2      3086336  1009858           152           2395              93
3      1090555  4388892            50            295              10
4       307789  4761599           161            664             104
...        ...      ...           ...            ...             ...
24995  4177982  2652529           151           7999               8
24996   305271  1110562           430            435            1172
24997  1522576  4630277           629           7998               1
24998  1229957   349493           696           7999              41
24999   690509   534148           103           3070               9

[25000 rows x 5 columns]

In [114]:
out_degrees_dest = G_complete_train_data.out_degree(combined_data_with_attributes['To'])
lst_1 = list(out_degrees_dest)
type(lst_1[0])
second_values_1 = [t[1] for t in lst_1]
second_values_1
combined_data_with_attributes['OutDegree_dest'] = second_values_1
combined_data_with_attributes

Form       To  src_Indegree  src_Outdegree  In_Degree_dest  \
0      4862306  3513873           291           7998              11   
1       629760  4763045           144           1351              23   
2      3086336  1009858           152           2395              93   
3      1090555  4388892            50            295              10   
4       307789  4761599           161            664             104   
...        ...      ...           ...            ...             ...   
24995  4177982  2652529           151           7999               8   
24996   305271  1110562           430            435            1172   
24997  1522576  4630277           629           7998               1   
24998  1229957   349493           696           7999              41   
24999   690509   534148           103           3070               9   

       OutDegree_dest  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
24995               0  
24996             481  
24997               0  
24998               0  
24999               0  

[25000 rows x 6 columns]

In [116]:
med_in_1 = combined_data_with_attributes['In_Degree_dest'].mean()
med_out_2 = combined_data_with_attributes['OutDegree_dest'].mean()
print(med_in_1)
print(med_out_2)

105.83592
261.19308


In [117]:
zero_count = (combined_data_with_attributes['OutDegree_dest'] == 0).sum()
print(zero_count)

22051


In [118]:
zero_count_In_Degree_dest = (combined_data_with_attributes['In_Degree_dest'] == 0).sum()
print(zero_count_In_Degree_dest)

0


##So here i can make one conclusion (for report) that , when i am calculating outdegree of destination node many nodes have 0 zero values precisely 22051 , intially i thought my code had a bug i went back to my toy example to cross check but code was working fine so i realized may be problem is with dataset, remember that unique person in our data is around 25 lacs and following list is given for 20,000 nodes so it makes sense for many outdegree of dest to be zero bcoz we dont know to whom they are following 

# Above code if for positive edge , now we will perfom same for negative edge

In [119]:
combined_data_with_attributes_negative_sample = pd.merge(df_false_ed, src_indegree_df , left_on='Form', right_on='Node', how='left')
combined_data_with_attributes_negative_sample = pd.merge(combined_data_with_attributes_negative_sample, src_outdegree_df, left_on='Form', right_on='Node', how='left')
                    

In [120]:
combined_data_with_attributes_negative_sample

Form       To     Node_x  src_Indegree     Node_y  src_Outdegree
0      1412960   159984  1412960.0           1.0  1412960.0            0.0
1       450019  3048705   450019.0           1.0   450019.0            0.0
2      2341943  2879266  2341943.0           2.0  2341943.0            0.0
3      3936197  4732550  3936197.0          12.0  3936197.0            0.0
4       258980   970828   258980.0           1.0   258980.0            0.0
...        ...      ...        ...           ...        ...            ...
24995   918964   186086        NaN           NaN        NaN            NaN
24996  2561631  2260842  2561631.0           1.0  2561631.0            0.0
24997  4741601  4778408  4741601.0           1.0  4741601.0            0.0
24998  1692402  2786562        NaN           NaN        NaN            NaN
24999  2931013  1568451        NaN           NaN        NaN            NaN

[25000 rows x 6 columns]

In [122]:
columns_to_drop = ['Node_x', 'Node_y']
combined_data_with_attributes_negative_sample = combined_data_with_attributes_negative_sample.drop(columns=columns_to_drop)

In [123]:
combined_data_with_attributes_negative_sample

Form       To  src_Indegree  src_Outdegree
0      1412960   159984           1.0            0.0
1       450019  3048705           1.0            0.0
2      2341943  2879266           2.0            0.0
3      3936197  4732550          12.0            0.0
4       258980   970828           1.0            0.0
...        ...      ...           ...            ...
24995   918964   186086           NaN            NaN
24996  2561631  2260842           1.0            0.0
24997  4741601  4778408           1.0            0.0
24998  1692402  2786562           NaN            NaN
24999  2931013  1568451           NaN            NaN

[25000 rows x 4 columns]

In [125]:
nan_count = combined_data_with_attributes_negative_sample['src_Indegree'].isna().sum()
print(nan_count)

12283


In [126]:
combined_data_with_attributes_negative_sample['src_Indegree'] = combined_data_with_attributes_negative_sample['src_Indegree'].fillna(0)
combined_data_with_attributes_negative_sample

Form       To  src_Indegree  src_Outdegree
0      1412960   159984           1.0            0.0
1       450019  3048705           1.0            0.0
2      2341943  2879266           2.0            0.0
3      3936197  4732550          12.0            0.0
4       258980   970828           1.0            0.0
...        ...      ...           ...            ...
24995   918964   186086           0.0            NaN
24996  2561631  2260842           1.0            0.0
24997  4741601  4778408           1.0            0.0
24998  1692402  2786562           0.0            NaN
24999  2931013  1568451           0.0            NaN

[25000 rows x 4 columns]

In [127]:
zero_count_src_Indegree = (combined_data_with_attributes_negative_sample['src_Indegree'] == 0).sum()
print(zero_count_src_Indegree)

12283


In [128]:
combined_data_with_attributes_negative_sample['src_Outdegree'] = combined_data_with_attributes_negative_sample['src_Outdegree'].fillna(0)
combined_data_with_attributes_negative_sample

Form       To  src_Indegree  src_Outdegree
0      1412960   159984           1.0            0.0
1       450019  3048705           1.0            0.0
2      2341943  2879266           2.0            0.0
3      3936197  4732550          12.0            0.0
4       258980   970828           1.0            0.0
...        ...      ...           ...            ...
24995   918964   186086           0.0            0.0
24996  2561631  2260842           1.0            0.0
24997  4741601  4778408           1.0            0.0
24998  1692402  2786562           0.0            0.0
24999  2931013  1568451           0.0            0.0

[25000 rows x 4 columns]

In [129]:
combined_data_with_attributes_negative_sample['src_Indegree'] = combined_data_with_attributes_negative_sample['src_Indegree'].astype(int)
combined_data_with_attributes_negative_sample['src_Outdegree'] = combined_data_with_attributes_negative_sample['src_Outdegree'].astype(int)
combined_data_with_attributes_negative_sample

Form       To  src_Indegree  src_Outdegree
0      1412960   159984             1              0
1       450019  3048705             1              0
2      2341943  2879266             2              0
3      3936197  4732550            12              0
4       258980   970828             1              0
...        ...      ...           ...            ...
24995   918964   186086             0              0
24996  2561631  2260842             1              0
24997  4741601  4778408             1              0
24998  1692402  2786562             0              0
24999  2931013  1568451             0              0

[25000 rows x 4 columns]

In [134]:
#in_degrees_dest =  G_complete_train_data.in_degree(combined_data_with_attributes_negative_sample['To'])
#lst = list(in_degrees_dest)
#type(lst[0])
#second_values = [t[1] for t in lst]
#second_values
#combined_data_with_attributes_negative_sample['In_Degree_dest'] = second_values
#combined_data_with_attributes_negative_sample

ValueError: Length of values (12633) does not match length of index (25000)

In [135]:
combined_data_with_attributes_negative_sample['In_Degree_dest'] = 0
combined_data_with_attributes_negative_sample

Form       To  src_Indegree  src_Outdegree  In_Degree_dest
0      1412960   159984             1              0               0
1       450019  3048705             1              0               0
2      2341943  2879266             2              0               0
3      3936197  4732550            12              0               0
4       258980   970828             1              0               0
...        ...      ...           ...            ...             ...
24995   918964   186086             0              0               0
24996  2561631  2260842             1              0               0
24997  4741601  4778408             1              0               0
24998  1692402  2786562             0              0               0
24999  2931013  1568451             0              0               0

[25000 rows x 5 columns]

In [136]:
combined_data_with_attributes_negative_sample['OutDegree_dest'] = 0
combined_data_with_attributes_negative_sample

Form       To  src_Indegree  src_Outdegree  In_Degree_dest  \
0      1412960   159984             1              0               0   
1       450019  3048705             1              0               0   
2      2341943  2879266             2              0               0   
3      3936197  4732550            12              0               0   
4       258980   970828             1              0               0   
...        ...      ...           ...            ...             ...   
24995   918964   186086             0              0               0   
24996  2561631  2260842             1              0               0   
24997  4741601  4778408             1              0               0   
24998  1692402  2786562             0              0               0   
24999  2931013  1568451             0              0               0   

       OutDegree_dest  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
24995               0  
24996               0  
24997               0  
24998               0  
24999               0  

[25000 rows x 6 columns]

In [137]:
combined_data_with_attributes_negative_sample['y'] = 0
combined_data_with_attributes_negative_sample

Form       To  src_Indegree  src_Outdegree  In_Degree_dest  \
0      1412960   159984             1              0               0   
1       450019  3048705             1              0               0   
2      2341943  2879266             2              0               0   
3      3936197  4732550            12              0               0   
4       258980   970828             1              0               0   
...        ...      ...           ...            ...             ...   
24995   918964   186086             0              0               0   
24996  2561631  2260842             1              0               0   
24997  4741601  4778408             1              0               0   
24998  1692402  2786562             0              0               0   
24999  2931013  1568451             0              0               0   

       OutDegree_dest  y  
0                   0  0  
1                   0  0  
2                   0  0  
3                   0  0  
4                   0  0  
...               ... ..  
24995               0  0  
24996               0  0  
24997               0  0  
24998               0  0  
24999               0  0  

[25000 rows x 7 columns]

In [138]:
combined_data_with_attributes['y'] = 1
combined_data_with_attributes

Form       To  src_Indegree  src_Outdegree  In_Degree_dest  \
0      4862306  3513873           291           7998              11   
1       629760  4763045           144           1351              23   
2      3086336  1009858           152           2395              93   
3      1090555  4388892            50            295              10   
4       307789  4761599           161            664             104   
...        ...      ...           ...            ...             ...   
24995  4177982  2652529           151           7999               8   
24996   305271  1110562           430            435            1172   
24997  1522576  4630277           629           7998               1   
24998  1229957   349493           696           7999              41   
24999   690509   534148           103           3070               9   

       OutDegree_dest  y  
0                   0  1  
1                   0  1  
2                   0  1  
3                   0  1  
4                   0  1  
...               ... ..  
24995               0  1  
24996             481  1  
24997               0  1  
24998               0  1  
24999               0  1  

[25000 rows x 7 columns]

In [139]:
comb_train_data = pd.concat([combined_data_with_attributes, combined_data_with_attributes_negative_sample], ignore_index=True)

In [140]:
comb_train_data

Form       To  src_Indegree  src_Outdegree  In_Degree_dest  \
0      4862306  3513873           291           7998              11   
1       629760  4763045           144           1351              23   
2      3086336  1009858           152           2395              93   
3      1090555  4388892            50            295              10   
4       307789  4761599           161            664             104   
...        ...      ...           ...            ...             ...   
49995   918964   186086             0              0               0   
49996  2561631  2260842             1              0               0   
49997  4741601  4778408             1              0               0   
49998  1692402  2786562             0              0               0   
49999  2931013  1568451             0              0               0   

       OutDegree_dest  y  
0                   0  1  
1                   0  1  
2                   0  1  
3                   0  1  
4                   0  1  
...               ... ..  
49995               0  0  
49996               0  0  
49997               0  0  
49998               0  0  
49999               0  0  

[50000 rows x 7 columns]

In [144]:
csv_file_path = 'sample_dataframe.csv'
comb_train_data.to_csv(csv_file_path, index=False)

In [150]:
#Adding Same features to the kaggle test

In [151]:
tf = pd.read_csv('real_test_data.csv')
tf

Id     Form       To
0        1  3360982  4457271
1        2  4761876  4698439
2        3  4198430  3615486
3        4  2945770   747948
4        5  3950088  3360335
...    ...      ...      ...
1995  1996  3116985  2415415
1996  1997   265459   842701
1997  1998  2547139  4332939
1998  1999  2366264  1815056
1999  2000  2735577  1232247

[2000 rows x 3 columns]

In [152]:
tf = tf.drop(columns='Id')
tf

Form       To
0     3360982  4457271
1     4761876  4698439
2     4198430  3615486
3     2945770   747948
4     3950088  3360335
...       ...      ...
1995  3116985  2415415
1996   265459   842701
1997  2547139  4332939
1998  2366264  1815056
1999  2735577  1232247

[2000 rows x 2 columns]

In [153]:
tf = pd.merge(tf, src_indegree_df , left_on='Form', right_on='Node', how='left')
tf = pd.merge(tf, src_outdegree_df, left_on='Form', right_on='Node', how='left')
tf
                    

Form       To   Node_x  src_Indegree   Node_y  src_Outdegree
0     3360982  4457271  3360982           102  3360982           1964
1     4761876  4698439  4761876            68  4761876            123
2     4198430  3615486  4198430           217  4198430           6996
3     2945770   747948  2945770           210  2945770            289
4     3950088  3360335  3950088           149  3950088           1309
...       ...      ...      ...           ...      ...            ...
1995  3116985  2415415  3116985           105  3116985            449
1996   265459   842701   265459            95   265459           3947
1997  2547139  4332939  2547139            48  2547139            131
1998  2366264  1815056  2366264            10  2366264            107
1999  2735577  1232247  2735577            17  2735577            544

[2000 rows x 6 columns]

In [154]:
columns_to_drop = ['Node_x', 'Node_y']
tf = tf.drop(columns=columns_to_drop)
tf

Form       To  src_Indegree  src_Outdegree
0     3360982  4457271           102           1964
1     4761876  4698439            68            123
2     4198430  3615486           217           6996
3     2945770   747948           210            289
4     3950088  3360335           149           1309
...       ...      ...           ...            ...
1995  3116985  2415415           105            449
1996   265459   842701            95           3947
1997  2547139  4332939            48            131
1998  2366264  1815056            10            107
1999  2735577  1232247            17            544

[2000 rows x 4 columns]

In [156]:
tf['In_Degree_dest'] = 0
tf

Form       To  src_Indegree  src_Outdegree  In_Degree_dest
0     3360982  4457271           102           1964               0
1     4761876  4698439            68            123               0
2     4198430  3615486           217           6996               0
3     2945770   747948           210            289               0
4     3950088  3360335           149           1309               0
...       ...      ...           ...            ...             ...
1995  3116985  2415415           105            449               0
1996   265459   842701            95           3947               0
1997  2547139  4332939            48            131               0
1998  2366264  1815056            10            107               0
1999  2735577  1232247            17            544               0

[2000 rows x 5 columns]

In [157]:
tf['OutDegree_dest'] = 0
tf

Form       To  src_Indegree  src_Outdegree  In_Degree_dest  \
0     3360982  4457271           102           1964               0   
1     4761876  4698439            68            123               0   
2     4198430  3615486           217           6996               0   
3     2945770   747948           210            289               0   
4     3950088  3360335           149           1309               0   
...       ...      ...           ...            ...             ...   
1995  3116985  2415415           105            449               0   
1996   265459   842701            95           3947               0   
1997  2547139  4332939            48            131               0   
1998  2366264  1815056            10            107               0   
1999  2735577  1232247            17            544               0   

      OutDegree_dest  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  
...              ...  
1995               0  
1996               0  
1997               0  
1998               0  
1999               0  

[2000 rows x 6 columns]

In [158]:
csv_file_path = 'test_kaggle_model_1.csv'
tf.to_csv(csv_file_path, index=False)

In [ ]:
csv_file_path = 'sample_dataframe.csv'
comb_train_data.to_csv(csv_file_path, index=False)